### DISCUSSION TITLE
Render custom tag at the end

### DISCUSSION THREAD

**Discussion by bfirems** [remarkjs/remark#1362] [Q&A]
I'm able to render my html at the end of say a p or li tag, however the issue arises when it generates more than one tag, it will render twice. Wondering if there is a way to render my html at the end, only once no matter what. I can't do it right after because it's a citation link that needs to be inside of the li (if it's the last thing rendered) to display properly.

**Comment by ChristianMurphy** [comment]
@bfirems I don't really understand your question.
1. Which remark library or libraries are you using?
2. What is "it" that renders twice? (The element, the whole page, something else)
3. Why does it matter if it renders twice?
4. Could you share a reproduction using https://stackblitz.com so others can see what you see?

**Reply by bfirems** [reply]
Thanks for your response!

1. react-markdown
2. "it" is "my html". What I want rendered after the markdown
3. It's a citation link, it looks weird and confusing if duplicated
4. https://stackblitz.com/edit/vitejs-vite-kgny55?file=src%2FApp.tsx

**Reply by ChristianMurphy** [reply]
> https://stackblitz.com/edit/vitejs-vite-kgny55?file=src%2FApp.tsx

I don't understand what you're showing me.
The item only renders once to HTML.
and the component is only called once for that render.
You can see this by adding a console log, and seeing the log get called once, as shown in the link below.
https://stackblitz.com/edit/vitejs-vite-okepw1?file=src%2FApp.tsx

> It's a citation link, it looks weird and confusing if duplicated

What do you mean by duplicated?
In your example you manually entered the citation twice.
Are you looking to create a plugin which looks for similar text and deduplicates it?
Or could you explain what you're issue is more clearly?

**Comment by bfirems** [comment]
Yes it only renders once to html, but it renders 2 li tags, one for ol because of the 1. and then another for the -, which renders an li under a ul. Since I'm overriding ul this isn't going to work. If you notice in my example it render two &nbspC which is just me simulating a citation.

What we are doing is rendering citation links as they pop up in the text, which could be anywhere in the markdown essentially. The 2 options I've found are either using remarkGfm and rendering citation tags (i.e. [^1]). So basically we would take our text and replace [789, 999, etc] with [^1][^2]. The problem with this solution is that if you don't have the footnote section it doesn't render. Our footnote section is a drawer than opens up once you click a citation. The other solution is just creating our own custom plugin. I haven't found good documentation on how to do that yet. Unless you have any other suggestions.

**Reply by ChristianMurphy** [reply]
> The 2 options I've found are either using remarkGfm and rendering citation tags (i.e. [^1]). So basically we would take our text and replace [789, 999, etc] with [^1][

Using standard markdown footnotes for citations would likely be ideal.
I'd recommend against creating yet another markdown flavor.

Extend the syntax of markdown means:

* Authors won’t be familiar with the syntax
* Content won’t work in other places (such as on GitHub)
* Defeating the purpose of markdown: being simple to author and looking like what it means

source: https://github.com/micromark/micromark?tab=readme-ov-file#extending-markdown

> The problem with this solution is that if you don't have the footnote section it doesn't render. 

I don't understand your comment, footnotes don't need a section to render, they're always added at the end.
And could be picked out by a plugin.

> The other solution is just creating our own custom plugin. I haven't found good documentation on how to do that yet. 

I'd recommend starting with https://unifiedjs.com/learn/ which introduces ASTs and plugins.

**Reply by bfirems** [reply]
This is an internal app, where our research team already wrote code to return markdown with citations in a specific format. We not only need to know the number of the citation, but the type as well. We are basically beholden to this format for now. 

As far as the footnotes, I'm talking about the sup tag. It won't render a sup unless we have a footnote markdown block at the bottom somewhere (The [1]: explanation). Basically what I want to do is override the sup tag with our Citation component, and not have a footnote section at the bottom. When you click the citation it opens a drawer with the footnote info.

Probably best if we can replace a block of text via regex with our citation component. I'll look at unified. Thank you.

**Reply by ChristianMurphy** [reply]
@bfirems this does and continues to feel a lot like an [XY problem](https://xyproblem.info/).
It would be really helpful to give information on what you are trying to solve as up front as possible, instead of asking about ideas of solutions that may or may not actually solve the problem you have.

> where our research team already wrote code to return markdown with citations in a specific format

If the code is already written and done, why is there a question on how to do it? 😅 

> We are basically beholden to this format for now.

Not necessarily, you could write a one off transform to convert the existing format to standard markdown, and move forward with that.

> Basically what I want to do is override the sup tag with our Citation component, and not have a footnote section at the bottom. When you click the citation it opens a drawer with the footnote info.

Remark plugins allow you to control when/where/if content is inserted, you could also keep it out of the page if wanted.
I'm speaking pretty generically here because I still don't have an example of an input markdown to output html you want.

> I'll look at unified.

Remark is part of unified https://unifiedjs.com/learn/guide/introduction-to-unified/
It would still be in remark or react-markdown.

**Reply by bfirems** [reply]
Sure I can explain the problem in greater detail. We use an external package to make LLM calls. That package is owned by another team, and the context prompt tells the LLM how to respond with citations. The format is 

```
#Some heading - some more text [Data: Type1(10, 11), Type2 (34, 35) some more text, another block of text [Data: Type1(99), Type2 (34)]". 
```

We want that turned into: 

```
<h1>Some heading</h1>
<ul>
    <li>some more text <Citation onClick="handeClick(10)">1</Citation><Citation onClick="handeClick(11)">2</Citation><Citation onClick="handeClick(34)">3</Citation><Citation onClick="handeClick(35)">4</Citation>
</ul>

<p>
    some more text, another block of text  <Citation onClick="handeClick(99)">5</Citation><Citation onClick="handeClick(34)">3</Citation>
</p>
```

We need to know Type1, 2 etc but that's beyond the scope of the markdown component.

**Reply by ChristianMurphy** [reply]
Clarifying a few things

Is this intentional?

```md
#Some
```

This isn't a markdown heading, a space is required, are you doing custom parsing to support this?

---

What is `[Data: ]`?
Does it also do something, it looks like in your example it can be un-closed? Does it have an HTML like auto-close ability?

---

For Type is space allowed between that and the parenthesis?
What does the number before the parenthesis mean?
Inside the parenthesis can there only be two numbers? Or can it have none, one, or many as well?

```md
Type1(10, 11), Type2 (34, 35)
```

---

> context prompt tells the LLM how to respond with citations

This is looking a bit less like citations are more like a directive system.
If this is LLM generated, could you ask the LLM to generate standard directives? https://github.com/remarkjs/remark-directive

**Reply by bfirems** [reply]
It's just example markdown, not to the tee exact, a space here and there are just typos. That's not really the point. The point is to take some regex in between [] and turn it into a custom react component with an onlick handler. Everything outside the brackets will be markdown and needs to be turned into html

**Reply by bfirems** [reply]
And no, like I said we can't control what the LLM prompt is, we have to deal with the response as is, that's just how they coded it.

**Reply by bfirems** [reply]
It is citations not a directive. Have you used Copilot or ChatGPT before? You know how they respond and give you citations to click on, where they are deriving that information from?

**Reply by ChristianMurphy** [reply]
> The point is to take some regex in between [] and turn it into a custom react component with an onlick handler. Everything outside the brackets will be markdown and needs to be turned into html

It is kinda the point.
If the model has an imprecise format, that regular expression will need a lot of error and edge case handling.
Tool usage

> And no, like I said we can't control what the LLM prompt is, we have to deal with the response as is, that's just how they coded it.

This feels like the root of the problem.
If you need to build a product off an LLM, you should either have some say over the prompt or a pre-built rendering library for it.

I understand this may not be the case for you right now.
But you're going down a rabbit hole of trying solving an organization coordination problem with code.

> It is citations not a directive. Have you used Copilot or ChatGPT before? You know how they respond and give you citations to click on, where they are deriving that information from?

I am very familiar with them.
What you are describing isn't ChatGPT's, Claude's, Mistral's, or Llama's citation format, it looks like something hand rolled.
All of the above are a lot stronger at generating XML or JSON than at generating custom markdown.

---

Again it is possible to parse custom markdown, regex alone may not be enough.
The syntax seems to overlap a bit with reference links in markdown.
You may need a parser extension in micromark https://github.com/micromark/micromark?tab=readme-ov-file#creating-a-micromark-extension (the parser remark builds on).

But it feels like trying to fight an organizational problem and dealing with output the LLMs are weaker at generating. With custom code, instead of fixing the output 🤷‍♂️

**Reply by bfirems** [reply]
It's not imprecise. We always get markdown, and for citations we always get brackets with the citations inside. Whether it's coming from an LLM or not is irrelevant. Forget I even said the word LLM, because it doesn't matter. We need to turn those brackets into a <Citation /> component with an onclick, and the rest needs to display as html like the control normally does. Whether that's how it should be returned to us or not is also irrelevant, because it's outside my control.

**Reply by ChristianMurphy** [reply]
> Whether that's how it should be returned to us or not is also irrelevant, because it's outside my control.

Fair enough.
I just wanted to be clear, there's an easy way, generate plain valid markdown.
And a hard way, creating a custom flavor of markdown.

You've chosen the hard way. 🤷‍♂️ 🙂 

Still doable, you'll need a parser https://github.com/micromark/micromark?tab=readme-ov-file#creating-a-micromark-extension
A way to take those tokens and convert them into an AST.

Add instructions of how to convert your custom node type back into HTML/React https://github.com/syntax-tree/mdast-util-to-hast?tab=readme-ov-file#fields-on-nodes

And some pretty thorough testing to make that production ready.

---

Making your own markdown flavor is pretty involved, but doable, feel free to as questions as you embark on the journey.
Good luck! 🍀
